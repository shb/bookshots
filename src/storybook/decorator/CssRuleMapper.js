export default class CssRuleMapper
{
  constructor (options) {
    this.stylesheet = undefined;
    Object.assign(this, options || {})
    this.rules = new Map;
  }

  set (selector, style) {
    this.rules.set(selector, style)
    this.refreshRules()
  }

  delete(selector) {
    this.rules.delete(selector)
    this.refreshRules()
  }

  refreshRules () {
    while (this.stylesheet.cssRules.length > 0) {
      this.stylesheet.deleteRule(0)
    }

    this.rules.forEach((style, selector) => {
      const rule = `${selector} ${style}`
      this.stylesheet.insertRule(rule, this.stylesheet.cssRules.length)
    })
  }
}