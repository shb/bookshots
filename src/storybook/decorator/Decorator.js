import Decoration from "./Decoration"
import StoryHelper from './StoryHelper'

export default class Decorator
{
  constructor (options) {
    this.mapper = undefined
    Object.assign(this, options || {})
    this.color = 'red'
    this.counters = []
  }

  flagStory (story) {
    const id = story.id
    this.mapper.set(this.getSelector(id), new Decoration('bullet', {
      color: this.color
    }))

    this.flagStoryParents (story)
  }

  unflagStory (story) {
    const id = story.id
    this.mapper.delete(this.getSelector(id))

    this.unflagStoryParents(story)
  }

  flagStoryParents (story) {
    const parents = StoryHelper.getParentIds(story)
    parents.forEach(pid => {
      const count = this.incrementCounter(pid)
      this.mapper.set(this.getSelector(pid), new Decoration('badge', {
        color: this.color,
        content: count
      }))
    })
  }

  unflagStoryParents (story) {
    const parents = StoryHelper.getParentIds(story)
    parents.forEach(pid => {
      const count = this.decrementCounter(pid)
      if (count <= 0) {
        this.mapper.delete(this.getSelector(pid))
      } else {
        this.mapper.set(this.getSelector(pid), new Decoration('badge', {
          color: this.color,
          content: count
        }))
      }
    })
  }

  incrementCounter (id) {
    if (!(id in this.counters)) this.counters[id] = 0
    this.counters[id] ++
    return this.counters[id]
  }

  decrementCounter (id) {
    if (!(id in this.counters)) return 0
    this.counters[id] --
    if (this.counters[id] < 0 ) this.counters[id] = 0
    return this.counters[id]
  }

  getSelector (id) {
    return `#${id}:after`
  }
}
