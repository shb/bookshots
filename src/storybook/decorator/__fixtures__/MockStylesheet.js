export default  class MockStylesheet
{
  constructor () {
    this.rules = []
  }

  get cssRules () {
    const mock = this
    return {
      get length () {
        return mock.rules.length
      }
    }
  }

  insertRule (rule, index) {
    this.rules.splice(index || 0, 0, rule)
  }

  deleteRule (index) {
    this.rules.splice(index, 1)
  }

  toString() {
    return this.rules.join("\n")
  }
}