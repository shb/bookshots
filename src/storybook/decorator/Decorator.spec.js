import Decorator from './Decorator'
import CssRuleMapper from './CssRuleMapper';
import MockStylesheet from './__fixtures__/MockStylesheet';

describe("Decorator", () =>
{
  let stylesheet
  let deco

  beforeEach(() => {
    stylesheet = new MockStylesheet
    const mapper = new CssRuleMapper({ stylesheet })
    deco = new Decorator({ mapper })
  })

  test("flagging one story", () => {
    const story = {
      id: 'button--simple',
      kind: "Button",
      name: 'simple'
    }
    deco.flagStory(story)
    expect(stylesheet.toString()).toMatchSnapshot()
  })

  test("flag two sibling stories", () => {
    const story1 = {
      id: 'button--simple',
      kind: "Button",
      name: 'simple'
    }
    const story2 = {
      id: 'button--not-so-simple',
      kind: "Button",
      name: 'not-so-simple'
    }
    deco.flagStory(story1)
    deco.flagStory(story2)
    expect(stylesheet.toString()).toMatchSnapshot()
  })

  test("unflagging a story", () => {
    const story1 = {
      id: 'button--simple',
      kind: "Button",
      name: 'simple'
    }
    deco.flagStory(story1)
    deco.unflagStory(story1)
    expect(stylesheet.toString()).toMatchSnapshot()
  })

  test("unflagging one of two sibling flagged stories", () => {
    const story1 = {
      id: 'button--simple',
      kind: "Button",
      name: 'simple'
    }
    const story2 = {
      id: 'button--not-so-simple',
      kind: "Button",
      name: 'not-so-simple'
    }
    deco.flagStory(story1)
    deco.flagStory(story2)
    deco.unflagStory(story1)
    expect(stylesheet.toString()).toMatchSnapshot()
  })
})
