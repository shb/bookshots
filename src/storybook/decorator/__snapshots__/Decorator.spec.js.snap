// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Decorator flag two sibling stories 1`] = `
"#button--simple:after {
    display: inline-block;
    content: \\"\\";
    background: red;
    width: 1em;
    height: 1em;
    border-radius: 1em;
    font-size: 85%;
    margin: 0 1ex;
    /* position: absolute; */
    right: 2em;
}
#button:after {
    display: inline-block;
    content: \\"2\\";
    background: red;
    min-width: 1em;
    border-radius: 10em;
    vertical-align: baseline;
    padding: 2px 1ex 1px;
    font-size: 85%;
    margin: 0 1ex;
    text-align: center;
    font-weight: bold;
    color: white;
    line-height: 1;
    /* position: absolute; */
    right: 2em;
}
#button--not-so-simple:after {
    display: inline-block;
    content: \\"\\";
    background: red;
    width: 1em;
    height: 1em;
    border-radius: 1em;
    font-size: 85%;
    margin: 0 1ex;
    /* position: absolute; */
    right: 2em;
}"
`;

exports[`Decorator flagging one story 1`] = `
"#button--simple:after {
    display: inline-block;
    content: \\"\\";
    background: red;
    width: 1em;
    height: 1em;
    border-radius: 1em;
    font-size: 85%;
    margin: 0 1ex;
    /* position: absolute; */
    right: 2em;
}
#button:after {
    display: inline-block;
    content: \\"1\\";
    background: red;
    min-width: 1em;
    border-radius: 10em;
    vertical-align: baseline;
    padding: 2px 1ex 1px;
    font-size: 85%;
    margin: 0 1ex;
    text-align: center;
    font-weight: bold;
    color: white;
    line-height: 1;
    /* position: absolute; */
    right: 2em;
}"
`;

exports[`Decorator unflagging a story 1`] = `""`;

exports[`Decorator unflagging one of two sibling flagged stories 1`] = `
"#button:after {
    display: inline-block;
    content: \\"1\\";
    background: red;
    min-width: 1em;
    border-radius: 10em;
    vertical-align: baseline;
    padding: 2px 1ex 1px;
    font-size: 85%;
    margin: 0 1ex;
    text-align: center;
    font-weight: bold;
    color: white;
    line-height: 1;
    /* position: absolute; */
    right: 2em;
}
#button--not-so-simple:after {
    display: inline-block;
    content: \\"\\";
    background: red;
    width: 1em;
    height: 1em;
    border-radius: 1em;
    font-size: 85%;
    margin: 0 1ex;
    /* position: absolute; */
    right: 2em;
}"
`;
